Modifications to compute_transects

vanroekel opened the pull request:
This makes a number of modifications

1) removes mfdataset xarray call for netcdf4
2) no longer need a history file (not saved in E3SM now)
3) ability to plot multiple transects with string list
4) ability to plot all transects with -n 'all'
5) saves output to netcdf file for easy concat and no need to recompute
transport

vanroekel commented:
@pwolfram here are the updates to the transect code.  Certainly lots of room for improvement, but it can compute across transects now and doesn't blow memory on the 18to6.

pwolfram commented:
@vanroekel, thanks for these updates!  I'm going to check it out and will get back to you

pwolfram commented on ocean/transects/compute_transects.py:
```diff
@@ -9,7 +9,6 @@
   ./compute_transects.py
```
I think you should add your name above at line 4 @vanroekel ...

pwolfram commented:
Testing on anvil with 

```
python compute_transects.py 
-k /home/ccsm-data/inputdata/ocn/mpas-o/oEC60to30v3/EC60to30v3_transect_masks.nc 
-m /home/ccsm-data/inputdata/ocn/mpas-o/oEC60to30v3/oEC60to30v3_60layer.170905.nc
-t '/lcrc/group/acme/jwolfe/acme_scratch/20171218.base.GMPAS-IAF.T62_oECv3.anvil/run/mpaso.hist.am.timeSeriesStatsMonthly.*.nc'
-n 'all' 
```
now...

pwolfram commented on ocean/transects/compute_transects.py:
```diff
@@ -21,6 +20,8 @@
 import numpy as np
 import matplotlib.pyplot as plt
 import xarray as xr
```
Should we just completely remove `xr` from this script for now?

I'm guessing you went with `netCDF4` over `xr` because it can open files more quickly, correct?

pwolfram commented on ocean/transects/compute_transects.py:
```diff
@@ -31,41 +32,111 @@ def get_mask_short_names(mask):
+    plt.savefig('transport' + transectList[j].replace(' ', '') +'.png')
+
+# Add calls to save transport and then can build up
+  ncid=Dataset('transport.nc',mode='w',clobber=True, format='NETCDF3_CLASSIC')
```
This would be the location where the output file name could be inserted, for instance.

vanroekel replied:
good idea.  I made the change

pwolfram commented on ocean/transects/compute_transects.py:
```diff
@@ -31,41 +32,111 @@ def get_mask_short_names(mask):
   return mask
 
-def compute_transport(output, timeavg, mesh, mask, name='Drake Passage'):
-  output = xr.open_mfdataset(output, concat_dim='Time', decode_times=False)
```
Are you sure you want to do this?  It seems like there is at least some utility in being able to specify the output file name.

pwolfram replied:
But, previously this was confusing because it was an *E3SM* output, not a script output so perhaps this is the way to go.

vanroekel replied:
given that E3SM doesn't put out the hist files, I think we should leave as is

pwolfram commented on ocean/transects/compute_transects.py:
```diff
@@ -31,41 +32,111 @@ def get_mask_short_names(mask):
-      required=True)
+#  parser.add_argument("-o", "--output_file_pattern", dest="output_filename_pattern",
+#      help="MPAS Filename pattern for output.", metavar="FILE",
+#      required=True)
```
I think you should just remove this text here if you decide to remove the output file name.

vanroekel replied:
done

pwolfram commented:
@vanroekel, how many years did you run this on this high-res?  I'm running into memory use of 35% for the EC60to30 and it has only processed about 40years.  I'm wondering if we may need to use xarray for the memory management capability.  

Can you please say more about what caused the memory blow-out for the 18to6?  I think I'm missing something here...

vanroekel commented:
I ran 15 years.  I'm really surprised at memory issues.  When processing times it opens one file at a time and only keeps no more than 16 reals per file.  So it is unclear to me why there are memory issues.  Any ideas?  does the 'vel' array get reallocated for every file?  vel and 't' are the only variables read more than once.

I can't remember the exact issues.  I seem to remember it was something about pulling a small variable that was uniform across files in an open_mfdataset, but can't remember for sure.

pwolfram commented on ocean/transects/compute_transects.py:
```diff
@@ -31,41 +32,111 @@ def get_mask_short_names(mask):
-  amask = amask.where(amask.transectEdgeGlobalIDs > 0, drop=True)
-  transectEdges = np.asarray(amask.transectEdgeGlobalIDs.values-1, dtype='i')
+  if name.lower() == 'all':
+    transectList = mask.shortNames[1:].values
```
Should be a `:` instead of `1:` because the first name needs computed too, in my case with the 60to30 this was Drake Passage.

vanroekel replied:
fixed.

pwolfram commented:
@vanroekel, I think this was a false alarm on my part.  The script completes cleanly without issue.  It just seemed somewhat weird to me that the memory usage was much higher than previously for the script, but the functionality here is greater so I think don't we shouldn't worry about it for now.  There are bigger priorities to address.  If we have issues with 60 yrs for the 18to6 then we should go back and fix it if that happens (seems unlikely).  Thanks for your really great work here! 👍 

Both the runs completed without issues.  I'm just rerunning to take care of that `1:` instead of `:` issue I listed in the comments.  It appears to have worked pretty great for the lions-share of the transects.  The script  takes about 8 hrs to complete for the EC60to30, but I think that is expected unless we get "clever" with file reads, which would require use of many nodes (probably not worth it).

I'll verify that the Drake Passage case is the same between the new and old scripts.  Once we confirm this I think we should merge with the only fix needed is the `:` change.

pwolfram commented on ocean/transects/compute_transects.py:
```diff
@@ -31,41 +32,111 @@ def get_mask_short_names(mask):
+    plt.legend(loc='best', frameon=False)
+    plt.ylabel('Transport (Sv)')
+    plt.xlabel('yrs')
+    plt.title('Transport for ' + name)
```
Note, under new option `all`, probably want to use specific title for this given case so each plot isn't labeled as "Transport for all".

vanroekel replied:
I think I have fixed this now too..  I have updated the plotting with all observational bounds available

pwolfram commented:
@vanroekel, great work!  Results are comparable between scripts (e.g., quick flip visually identical except for title):

### old:
![transportdrakepassage](https://user-images.githubusercontent.com/4295853/36214528-ea442838-1166-11e8-8c97-9c3dc7455679.png)

## new:
![transportdrakepassage](https://user-images.githubusercontent.com/4295853/36214465-cc7a776c-1166-11e8-8352-761698a9e3a4.png)

There were a few really small little things-- I'd put those in a new comment and I'll double-check them by inspection and then one of us should merge it. I'm happy to do so or you should do so if you want to alternatively.

vanroekel commented:
@pwolfram I think I addressed your comments.  I also updated the plotting to allow for multiple transects.  Please take a look when you have time.  Thanks!

pwolfram commented:
@vanroekel, I have a notebook that converts this information into a table with observations.  Would it be ok to include that information here?  Otherwise, I think this PR is ready to merge.  I can do a separate PR if would would prefer too.

pwolfram commented on ocean/transects/compute_transects.py:
```diff
@@ -99,44 +100,53 @@ def compute_transport(timeavg, mesh, mask, name='Drake Passage'):
+  obsDict = {'Drake Passage':[120,175],'Tasmania-Ant':[147,167],'Africa-Ant':None,'Antilles Inflow':[-23.1,-13.7], \
+          'Mona Passage':[-3.8,-1.4],'Windward Passage':[-7.2,-6.8],'Florida-Cuba':[30,33],'Florida-Bahamas':[30,33], \
+          'Indonesian Throughflow':[-21,-11],'Agulhas':[-90,-50],'Mozambique Channel':[-20,-8], \
+          'Bering Strait':[0.17,1.49],'Lancaster Sound':None,'Fram Strait':None,'Robeson Channel':None}
```
I accidentally doubled this work today, but this is a good check of both our methods.  The results are a little different.  Do you recall what references you used?  Thanks!

```
# observations (Positive values are north and eastward, see Ringler et al 2013)
obs = {'Drake Passage': {'mean': 134 , 'std dev': 14, 'ref': 'Nowlin and Klinck (1986)'}, 
       'Tasmania-Ant': {'mean': 157, 'std dev': 10, 'ref': 'Ganachaud and Wunsch (2000)'},
       'Africa-Ant': {'mean': 150, 'std dev': 30, 'ref': 'Estimated from Drake and Tasmania-Ant passages from Ganachaud and Wunsch (2000) http://onlinelibrary.wiley.com/doi/10.1029/2002JC001565/full'},
       'Indonesian Throughflow': {'mean': -15, 'std dev': 4, 'ref': 'Sprintall et al. (2009)'},
       'Agulhas': {'mean': -70, 'std dev': 20, 'ref': 'unclear from Ringler et al (2013)' },
       'Mozambique Channel': {'mean': -16, 'std dev': 13, 'ref': ' van der Werf et al. (2010)'},
       'Antilles Inflow': {'mean': -18.4, 'std dev': 4.7, 'ref': 'Johns et al. (2002) and Roemmich (1981)'},
       'Mona Passage': {'mean': -2.6, 'std dev': 1.2, 'ref': 'Johns et al. (2002) and Roemmich (1981).'},
       'Windward Passage': {'mean': -7.0, 'std dev': np.nan, 'ref': 'Johns et al. (2002) and Roemmich (1981).'},
       'Florida-Cuba': {'mean': 31, 'std dev': 1.5, 'ref': 'Johns et al. (2002) and Roemmich (1981).'},
       'Florida-Bahamas': {'mean': 31.5, 'std dev': 1.5, 'ref': 'Johns et al. (2002) and Roemmich (1981).'},
       'Bering Strait': {'mean': 0.8, 'std dev': 0.3, 'ref': 'Roach et al (1995) http://onlinelibrary.wiley.com/doi/10.1029/95JC01673/full'},
       'Lancaster Sound': {'mean': 0.8 , 'std dev': 0.3, 'ref': 'Prisenberg and Hamilton (2005) http://www.tandfonline.com/doi/abs/10.3137/ao.430101'},
       'Fram Strait': {'mean': -3, 'std dev': 3, 'ref': 'Schauer et al (2004) http://onlinelibrary.wiley.com/doi/10.1029/2003JC001823/full'},       
       'Robeson Channel': {'mean': -0.7 , 'std dev': .2 , 'ref': 'Maltrud and McClean (2005) https://www.sciencedirect.com/science/article/pii/S1463500303000684#BIB16'}
       }
```

vanroekel replied:
most of these values are from a table from @mark-petersen.  The Drake passage is altered based on new observations in a paper Mat M. sent around that showed the drake passage transport is 175 Sv.

pwolfram commented on ocean/transects/compute_transects.py:
```diff
@@ -32,12 +32,14 @@ def get_mask_short_names(mask):
   if name.lower() == 'all':
-    transectList = mask.shortNames[1:].values
+    transectList = mask.shortNames[:].values
+    condition = transectList != "Atlantic Transec"
```
I'm a little confused here, are there sometimes `"Atlantic Transec"` in the output?  I didn't see that in the file that was produced from the previous version of this script.  I think I'm missing something here.

vanroekel replied:
In the 18to6 transect file there is an Atlantic transect that didn't make sense to run over (it's for MOC) and do the computation, so I pulled it out.  But I don't think it is a big deal to leave it in.

pwolfram commented:
Notebook is "gisted" at https://gist.github.com/832f2eb1ca5b711b88f03afba5a93b53

vanroekel commented:
Yes I think adding your table functionality would be great here.  I would recommend altering Drake passage to include the new observations.  The reference is here
```
http://onlinelibrary.wiley.com/doi/10.1002/2016GL070319/full
```

pwolfram commented:
@vanroekel, I just pushed the notebook to build the paper table to your branch if you want to check it out.

vanroekel commented:
@pwolfram all looks good to me.  If you are ready, I think we can merge.

pwolfram commented on ocean/transects/compute_transects.py:
```diff
@@ -1,7 +1,7 @@
 """
 Name: compute_transects.py
-Author: Phillip J. Wolfram, Mark Petersen
+Author: Phillip J. Wolfram, Mark Petersen, Luke Van Roekel
```
Good catch

pwolfram commented:
@vanroekel, sounds good to me.

pwolfram commented:
@vanroekel, I'm not sure I can delete your branch, but we are least merged in now.  Thanks!
